addons/account/models/partner.py

```diff
@@ -167,11 +167,17 @@ def get_fiscal_position(self, partner_id, delivery_id=None):
         # This can be easily overridden to apply more complex fiscal rules
         PartnerObj = self.env['res.partner']
         partner = PartnerObj.browse(partner_id)
+        delivery = PartnerObj.browse(delivery_id)
 
-        # if no delivery use invoicing
-        if delivery_id:
-            delivery = PartnerObj.browse(delivery_id)
-        else:
+        company = self.env.company
+        eu_country_codes = self.env.ref('base.europe').country_ids.mapped('code')
+        has_vat = bool(company.vat) and bool(partner.vat)
+        if has_vat:
+            intra_eu = company.vat[:2] in eu_country_codes and partner.vat[:2] in eu_country_codes
+            vat_exclusion = company.vat[:2] == partner.vat[:2]
+
+        # If company and delivery have the same vat prefix, use invoicing
+        if not delivery or (has_vat and intra_eu and vat_exclusion):
```

Comment:
Some linters might complain about `intra_eu` and `vat_exclusion` possibly being undefined below
```suggestion
        intra_eu, vat_exclusion = False, False
        if company.vat and partner.vat:
            intra_eu = company.vat[:2] in eu_country_codes and partner.vat[:2] in eu_country_codes
            vat_exclusion = company.vat[:2] == partner.vat[:2]

        # If company and delivery have the same vat prefix, use invoicing
        if not delivery or (intra_eu and vat_exclusion):
```